Move EvenPodsSpread metadata computation logic as a PreFilter Plugin

Huang-Wei opened the pull request:
**What type of PR is this?**

/kind feature
/sig scheduling

**What this PR does / why we need it**:

Move EvenPodsSpread metadata computation logic as a PreFilter Plugin.

**Which issue(s) this PR fixes**:

Fixes #85925.

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

/cc @ahg-g

Huang-Wei committed:
Move EvenPodsSpread metadata computation logic as a PreFilter Plugin

Huang-Wei commented:
/hold

ahg-g commented on pkg/scheduler/framework/plugins/podtopologyspread/pod_topology_spread.go:
```diff
@@ -26,40 +26,128 @@ import (
+func getPodTopologySpreadMetadata(cycleState *framework.CycleState) (*predicates.PodTopologySpreadMetadata, error) {
+	c, err := cycleState.Read(preFilterStateKey)
+	if err != nil {
+		return nil, err
```
Read returns an error when the key is not found. CA may chose not to run PreFilter at all, and so the entry will not exist, we don't want to error on that.

Huang-Wei replied:
In that case, can they just bypass registering the plugin as a "PreFilter" extension? Then they won't reach here.

Huang-Wei replied:
I see your point now. It explains my concern at https://github.com/kubernetes/kubernetes/pull/86175#discussion_r357010185. Thanks.

Huang-Wei replied:
In this case, I need to throw an error when the meta is nil in Filter(). (Because autoscaler doesn't have logic implemented for the "else" block for even pods spread yet)

Will add the logic.

Huang-Wei replied:
Ah, hold on.

In EvenPodsSpread, it returns nil as a **legit** meta when it's unnecessary to proceed pre-computing, e.g., no spread constraints in the pod, etc. But for sure, the nil meta will still be written to `CycleState`.
So if CA chooses not to run PreFilter, then the state entry (wrapping the meta) will be nil, then we will get NotFound error. If we tolerate the NotFound error, and assign the meta to nil, it will get conflicted with its intentional semantics. (In that case, a nil meta will give green light for each Filter() call).

Huang-Wei replied:
Per offline talk with @ahg-g, raised a new commit to adjust the semantic of `nil` "PodTopologySpreadMetadata". It's intentional to keep it as a separate commit in case of reverting in the future.

ahg-g reviewed (commented):
Thanks, just one thing.

Huang-Wei committed:
Update semantics of EvenPodsSpread metadata object

- Now `nil` represents the meta is not set at all (in PreFilter phase)
- An empty `PodTopologySpreadMetadata` object denotes it's a legit meta and is set in PreFilter phase.

ahg-g commented on pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -1649,7 +1649,13 @@ func PodTopologySpreadPredicate(pod *v1.Pod, meta *PodTopologySpreadMetadata, no
-	if meta == nil || len(meta.tpPairToMatchNum) == 0 || len(meta.constraints) == 0 {
+	// nil meta is illegal.
+	if meta == nil {
+		return false, nil, errors.New("metadata not pre-computed for PodTopologySpreadPredicate")
```
We need a todo here to implement this, right?

Huang-Wei replied:
done.

ahg-g commented on pkg/scheduler/framework/plugins/podtopologyspread/pod_topology_spread.go:
```diff
@@ -118,11 +121,9 @@ func (pl *PodTopologySpread) RemovePod(ctx context.Context, cycleState *framewor
-	// It's possible that meta is set to nil intentionally.
-	if c == nil {
+		// The metadata wasn't pre-computed in prefilter. We ignore the error for now since
+		// we are able to handle that by computing it again (e.g. in Filter()).
```
is this actually true for pod topology spread?

Huang-Wei replied:
The nil meta will be caught in the Filter, so yes, it should be good.

ahg-g replied:
sure, but we are not computing it again now.

ahg-g commented:
/lgtm

Huang-Wei commented:
/hold cancel

Huang-Wei commented:
rebased

ahg-g commented:
/lgtm

ahg-g commented:
/lgtm

ahg-g commented:
/retest

ahg-g commented:
/retest

Huang-Wei commented:
/retest

ahg-g commented:
/retest

ahg-g commented:
/retest
